crates/bevy_ecs/src/event/mod.rs

```diff
@@ -1,39 +1,351 @@
+///
+/// The [`Event`] trait should generally be derived:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #
+/// #[derive(Event)]
+/// struct Speak {
+///     message: String,
+/// }
+/// ```
+///
+/// An [`Observer`](crate::observer::Observer) can then be added to watch for this event type:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #
+/// # #[derive(Event)]
+/// # struct Speak {
+/// #     message: String,
+/// # }
+/// #
+/// # let mut world = World::new();
+/// #
+/// world.add_observer(|speak: On<Speak>| {
+///     println!("{}", speak.message);
+/// });
+/// ```
+///
+/// The event can be triggered on the [`World`] using the [`trigger`](World::trigger) method:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #
+/// # #[derive(Event)]
+/// # struct Speak {
+/// #     message: String,
+/// # }
+/// #
+/// # let mut world = World::new();
+/// #
+/// # world.add_observer(|speak: On<Speak>| {
+/// #     println!("{}", speak.message);
+/// # });
+/// #
+/// # world.flush();
+/// #
+/// world.trigger(Speak {
+///     message: "Hello!".to_string(),
+/// });
+/// ```
+///
+/// # Triggers
+///
+/// Every [`Event`] has an associated [`Trigger`] implementation (via [`Event::Trigger`]), which defines which observers will run,
+/// what data will be passed to them, and the order they will be run in. Unless you are an internals developer or you have very specific
+/// needs, you don't need to worry too much about [`Trigger`]. When you derive [`Event`] (or a more specific event trait like [`EntityEvent`]),
+/// a [`Trigger`] will be provided for you.
+///
+/// The [`Event`] derive defaults [`Event::Trigger`] to [`GlobalTrigger`], which will run all observers that watch for the [`Event`].
+///
+/// # Entity Events
+///
+/// For events that "target" a specific [`Entity`], see [`EntityEvent`].
+#[diagnostic::on_unimplemented(
+    message = "`{Self}` is not an `Event`",
+    label = "invalid `Event`",
+    note = "consider annotating `{Self}` with `#[derive(Event)]`"
+)]
+pub trait Event: Send + Sync + Sized + 'static {
+    /// Defines which observers will run, what data will be passed to them, and the order they will be run in. See [`Trigger`] for more info.
+    type Trigger<'a>: Trigger<Self>;
+}
+
+/// An [`EntityEvent`] is an [`Event`] that is triggered for a specific [`EntityEvent::event_target`] entity:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # let mut world = World::default();
+/// # let entity = world.spawn_empty().id();
+/// #[derive(EntityEvent)]
+/// struct Explode {
+///     entity: Entity,
+/// }
+///
+/// world.add_observer(|event: On<Explode>, mut commands: Commands| {
+///     println!("Entity {} goes BOOM!", event.entity);
+///     commands.entity(event.entity).despawn();
+/// });
+///
+/// world.trigger(Explode { entity });
+/// ```
+///
+/// [`EntityEvent`] will set [`EntityEvent::event_target`] automatically for named structs with an `entity` field name (as seen above). It also works for tuple structs
```

Comment:
Future work: surely we could scan the fields in the macro and then set it to the right field if only one `Entity` field exists?

Reply:
We can, but "scanning for types" in macros introduces a new avenue for incorrectness: type aliases, same type name but mismatched type, etc. Macros don't really operate on the type level in this case, and pretending they do introduces jank / weirdness.